src/schemes/boundary/open_boundary/system.jl

```diff
@@ -0,0 +1,476 @@
+        v_neighbor_ref = reference_velocity(neighbor_position, t)
+
+        # Determine characteristic variables
+        density_term = -sound_speed^2 * (rho_b - rho_ref)
+        pressure_term = p_b - p_ref
+        velocity_term = rho_b * sound_speed * (dot(v_b - v_neighbor_ref, flow_direction))
+
+        kernel_ = smoothing_kernel(neighbor_system, distance)
+
+        characteristics[1, particle] += (density_term + pressure_term) * kernel_
+        characteristics[2, particle] += (velocity_term + pressure_term) * kernel_
+        characteristics[3, particle] += (-velocity_term + pressure_term) * kernel_
+
+        volume[particle] += kernel_
+    end
+
+    return system
+end
+
+@inline function prescribe_conditions!(characteristics, particle, ::OutFlow)
+    # J3 is prescribed (i.e. determined from the exterior of the domain).
+    # J1 and J2 is transimtted from the domain interior.
+    characteristics[3, particle] = zero(eltype(characteristics))
+
+    return characteristics
+end
+
+@inline function prescribe_conditions!(characteristics, particle, ::InFlow)
+    # Allow only J3 to propagate upstream to the boundary
+    characteristics[1, particle] = zero(eltype(characteristics))
+    characteristics[2, particle] = zero(eltype(characteristics))
+
+    return characteristics
+end
+
+@inline function update_quantities!(system::OpenBoundarySPHSystem, v, u, t)
+    (; density, pressure, characteristics, flow_direction, sound_speed,
+    reference_velocity, reference_pressure, reference_density) = system
+
+    @threaded for particle in each_moving_particle(system)
+        particle_position = current_coords(u, system, particle)
+
+        J1 = characteristics[1, particle]
+        J2 = characteristics[2, particle]
+        J3 = characteristics[3, particle]
+
+        rho_ref = reference_density(particle_position, t)
+        density[particle] = rho_ref + ((-J1 + 0.5 * (J2 + J3)) / sound_speed^2)
+
+        p_ref = reference_pressure(particle_position, t)
+        pressure[particle] = p_ref + 0.5 * (J2 + J3)
+
+        v_ref = reference_velocity(particle_position, t)
+        rho = density[particle]
+        v_ = v_ref + ((J2 - J3) / (2 * sound_speed * rho)) * flow_direction
+
+        for dim in 1:ndims(system)
+            v[dim, particle] = v_[dim]
+        end
+    end
+
+    return system
+end
+
+function check_domain!(system, v, u, v_ode, u_ode, semi)
+    (; boundary_zone, fluid_system) = system
+
+    u_fluid = wrap_u(u_ode, fluid_system, semi)
+    v_fluid = wrap_v(v_ode, fluid_system, semi)
+
+    neighborhood_search = get_neighborhood_search(system, fluid_system, semi)
+
+    for particle in each_moving_particle(system)
+        particle_coords = current_coords(u, system, particle)
+
+        # Check if boundary particle is outside the boundary zone
+        if !is_in_boundary_zone(boundary_zone, particle_coords)
+            transform_particle!(system, fluid_system, boundary_zone, particle,
```

Comment:
How about `convert_particle!`? This also avoids confusion with `transfer_particle!`.